### ISSUE TITLE
Add conditional PageBreak API

### ISSUE THREAD

**Issue by marcmognol** [QuestPDF/QuestPDF#706]
**Is your feature request related to a problem? Please describe.**
I would like to have the possibility to add a Page break in a document if the current page of my document is Front (odd) or Back (even).
Because I need to be able to start a new paragraph or a new page on the Front page.

**Describe the solution you'd like**
Add new API to specify when PageBreak should occurs:

`PageBreak(bool condition)` but we need to have the current page number.

or something like:

`PageBreakOnFront()` or `PageBreakOnBack()`

**Describe alternatives you've considered**
Thanks to this awesome library, I was able to reproduce this behavior with Dynamic component with the following:

```
public class ConditionalPageBreak : IDynamicComponent<int>
{
    public int State { get; set; }

    public DynamicComponentComposeResult Compose(DynamicContext context)
    {
        var content = context.CreateElement(element =>
        {
            var isEven = context.PageNumber % 2 == 0;
            element.ShowIf(isEven)
                   .ExtendVertical();
        });

        return new DynamicComponentComposeResult
        {
            Content = content,
            HasMoreContent = false,
        };
    }
}
```

Which is used like:

```
            page.Content()
                .Column(c =>
                {
                    c.Item().Dynamic(new ConditionalPageBreak());
                    c.Item().Text("Should appears on Front page").FontSize(38);
                });
```

Thank you!

**Comment by MarcinZiabek** [comment]
@burkaysucu FYI

I was thinking about a new API called `ShowWhen` or possibly an overload for `ShowIf`. This API could provide context similar to the `DynamicComponent` element, e.g., current page number, total number of pages, etc. Then, it expects a predicate that decides whether its content should be visible.

For example:

```sharp
page.Content().Column(column =>
{
    column.Spacing(10);

    foreach (var s in Enumerable.Range(0, 10))
    {
        foreach (var i in Enumerable.Range(0, Random.Shared.Next(10, 50)))
        {
            column
                .Item()
                .Height(40)
                .Width(150)
                .Background(Colors.Grey.Lighten3)
                .Text($"{s} - {i}");
        }   
        
        column.Item().PageBreak();
        column.Item().ShowWhen(x => x.PageNumber % 2 == 0).PageBreak(); // <- usage
    }
});
```

Produces:
 
[QuestPDF - ShowWhen.pdf](https://github.com/QuestPDF/QuestPDF/files/13354404/QuestPDF.-.ShowWhen.pdf)

Questions: 
1) What else could be provided in the context? 
2) What are potential other usages? (I would to avoid introducing features with very limited scope)

**Comment by burkaysucu** [comment]
Hey @MarcinZiabek ,

I just tried @marcmognol approach and it works (btw, thank you @marcmognol). Note that what I wanted in #675 is adding an empty page "after" a chapter with an odd number of pages. This approach adds an empty page "before" the next chapter if a chapter has an odd number of pages. When we want to start over the page numbers from 1 for each chapter, it needs a lot of extra work to align the displayed page numbers.

Your new proposal looks good and would solve #675. I was walking around the source code and was thinking about something similar, exposing the current page number as in your example. However, I think giving it in that `ShowWhen` context is very limiting. Would not it be more useful if one could reach the PageNumber value in the container? A few use cases:
- display the page number in the content area
- style the content depending on the page number
- collect the chapter page counts easily for preamble (alternative is using a dynamic component and running some calculations. we are doing it, and it is confusing.)

**Comment by MarcinZiabek** [comment]
You are right - having full context in the container would be more helpful. However, it significantly changes the process - such content cannot be easily built before document generation but needs to be executed during the generation process (as part of the layout algorithm). This, in turn, requires entire machinery introduced specifically for DynamicComponent. Ultimately, it would act as a minimal API for the DynamicComponent (with all negative performance implications).

What do you think?

**Comment by burkaysucu** [comment]
Now I am lost:) As said earlier, I am not very familiar with the internals of the library. You know it the best. Nevertheless, I think exposing only the page number in a readonly manner would not break things. Let's say we have an oracle who correctly guesses the current page number in any container. Then a user could use that number in any way he/she wants. I guess the build process would not break in such a case :)

**Comment by MarcinZiabek** [comment]
It is just a matter of complexity and performance 😀

Anyway, thank you for providing the list of potential issues that could accomplished with the new API. I am still trying to figure out their usefulness. What actual real-world requirements could be solved?

**Comment by burkaysucu** [comment]
As said, I am fine with your early suggestion `ShowWhen`. Of course it is your call :)

Some related and interesting reads:

https://en.wikipedia.org/wiki/Recto_and_verso
"By book publishing convention, the first page of a book, and sometimes of each section and chapter of a book, is a recto page,[[5]](https://en.wikipedia.org/wiki/Recto_and_verso#cite_note-5) and hence all recto pages will have odd numbers and all verso pages will have even numbers.[[6]](https://en.wikipedia.org/wiki/Recto_and_verso#cite_note-6)[[7]](https://en.wikipedia.org/wiki/Recto_and_verso#cite_note-7)"

https://word.tips.net/T003843_Making_Sure_a_Document_Always_Has_an_Even_Number_of_Pages.html
"Zelda would like to be able to automatically add a blank page at the end of a Word document when that document contains an odd number of pages. She would like this to happen because she produces many separate documents that are combined in Adobe and created as PDFs for double-sided printing."

https://en.wikipedia.org/wiki/Intentionally_blank_page
Examples section lists interesting uses.

https://www.todayifoundout.com/index.php/2016/11/pages-intentionally-left-blank-say/

**Comment by marcmognol** [comment]
Hi @MarcinZiabek, @burkaysucu 

Thank you both of you for the consideration of this proposal.

I do like the ShowWhen proposal.

Thanks for all.

Marc

**Comment by MarcinZiabek** [comment]
@burkaysucu Thank you for providing all examples! Funny thing, once I know about intentionally blank pages, I see them everywhere!

I will do my best to release this improvement in the next release, in November or early December. I would like to experiment more with potential layout instabilities.

**Comment by burkaysucu** [comment]
Hi Marcin,

Sounds good. Thank you for all the effort!

**Comment by burkaysucu** [comment]
Hello Marcin,

Any updates on this issue?

**Comment by girlpunk** [comment]
I belive the `ShowIf` function has now been released?

**Comment by burkaysucu** [comment]
@girlpunk Thank you for the info. Just checked the release notes and found `ShowIf` listed for 2023.12.0. We will upgrade and test it asap.

**Comment by JeremyVm** [comment]
@MarcinZiabek why does this ShowIf component not contain the AvailableSize like the DynamicComponent?

I want to show the Item if the AvailableSize.Height > x, if not then I want a page break.

**Comment by burkaysucu** [comment]
@JeremyVm Does EnsureSpace work for you? https://www.questpdf.com/api-reference/ensure-space.html

**Comment by JeremyVm** [comment]
@burkaysucu No, I'm using that but doesn't help.

**Comment by MarcinZiabek** [comment]
Based on user feedback, the ShowIf API has been enhanced in the 2023.12.0 release with a new overload that provides access to the current page number and the total number of pages in the document.

Additionally, the library now offers finer control over page breaks with two recently improved APIs:
- [Prevent Page Break](https://www.questpdf.com/api-reference/prevent-page-break.html)
- [Ensure Space](https://www.questpdf.com/api-reference/ensure-space.html)

Since there has been no recent activity, I’m closing this issue. If you have a similar request in the future, feel free to open a new GitHub issue. Thanks for your feedback! ❤️
